File: frontend/pages/edit_class.py
import streamlit as st

from services.class_api import ClassApiError, update_class


WEEKDAY_OPTIONS = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday",
]

WEEKDAY_LABELS = {
    "mon": "Monday",
    "monday": "Monday",
    "tue": "Tuesday",
    "tues": "Tuesday",
    "tuesday": "Tuesday",
    "wed": "Wednesday",
    "weds": "Wednesday",
    "wednesday": "Wednesday",
    "thu": "Thursday",
    "thur": "Thursday",
    "thurs": "Thursday",
    "thursday": "Thursday",
    "fri": "Friday",
    "friday": "Friday",
    "sat": "Saturday",
    "saturday": "Saturday",
    "sun": "Sunday",
    "sunday": "Sunday",
}


def _safe_text(value: object, fallback: str = "") -> str:
    if value is None:
        return fallback
    text = str(value).strip()
    return text if text else fallback


def _normalize_schedule_day(value: object) -> str:
    normalized = _safe_text(value, fallback="")
    if not normalized:
        return ""
    return WEEKDAY_LABELS.get(normalized.lower(), normalized)


def _render_hero(class_name: str) -> None:
    st.markdown(
        f"""
        <div class='analysis-hero'>
            <h2 style='margin:0'>Edit Class</h2>
            <p class='analysis-subtitle'>Update details for {class_name} and keep schedules up to date.</p>
            <div class='analysis-chip-row'>
                <span class='analysis-chip'>Class Settings</span>
                <span class='analysis-chip'>Schedule Update</span>
                <span class='analysis-chip'>Save Changes</span>
            </div>
        </div>
        """,
        unsafe_allow_html=True,
    )


def edit_class_page() -> None:
    token = st.session_state.get("auth_token")

    cls = st.session_state.get("selected_class")
    if not cls:
        st.warning("No class selected")
        if st.button("Back to My Classes", key="edit_back_without_class"):
            st.session_state.page = "My Classes"
            st.rerun()
        return

    class_id = cls.get("id")
    if not isinstance(class_id, int):
        st.error("Selected class is missing a valid id.")
        if st.button("Back to My Classes", key="edit_back_invalid_class"):
            st.session_state.page = "My Classes"
            st.rerun()
        return

    class_name_default = _safe_text(cls.get("class_name"))
    grade_default = _safe_text(cls.get("grade_age_group"))
    description_default = _safe_text(cls.get("description"))

    selected_schedule = [
        normalized
        for day in (cls.get("schedule_days") or [])
        for normalized in [_normalize_schedule_day(day)]
        if normalized in WEEKDAY_OPTIONS
    ]

    _render_hero(class_name_default or "this class")

    st.markdown("<div class='card'>", unsafe_allow_html=True)
    with st.form("edit_class_form", clear_on_submit=False):
        class_name = st.text_input("Class Name", value=class_name_default, placeholder="e.g. Grade 4A")
        grade_age_group = st.text_input("Grade / Age Group", value=grade_default, placeholder="e.g. Grade 4")
        schedule_days = st.multiselect(
            "Schedule Days",
            options=WEEKDAY_OPTIONS,
            default=selected_schedule,
            placeholder="Pick one or more days",
        )
        description = st.text_area(
            "Description",
            value=description_default,
            placeholder="Optional class notes or focus areas.",
            height=100,
        )

        submitted = st.form_submit_button("Save Changes", type="primary")

    st.markdown("</div>", unsafe_allow_html=True)

    action_col1, action_col2 = st.columns([1.2, 4.8])
    with action_col1:
        if st.button("Cancel", key="edit_cancel", use_container_width=True):
            st.session_state.page = "class_detail"
            st.rerun()

    if submitted:
        if not class_name.strip() or not grade_age_group.strip() or not schedule_days:
            st.error("Class name, grade/age group, and at least one schedule day are required.")
            return

        try:
            updated = update_class(
                class_id=class_id,
                class_name=class_name.strip(),
                grade_age_group=grade_age_group.strip(),
                schedule_days=schedule_days,
                description=description.strip(),
                token=token,
            )
        except ClassApiError as exc:
            st.error(f"Failed to update class: {exc}")
            return

        st.session_state.selected_class = updated
        st.success("Class updated successfully.")
        st.session_state.page = "class_detail"
        st.rerun()

File: frontend/js/pages/edit-class.js
import { createShell, attachShellHandlers } from '../components/shell.js';
import { fetchClassById, updateClass } from '../api/classes.js';
import { showToast } from '../components/toast.js';

const DAY_OPTIONS = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];

export async function renderPage({ params } = {}) {
	const appElement = document.getElementById('app');
	if (!appElement) return;

	const classId = params?.id;
	if (!classId) {
		showToast('error', 'Class not found.');
		window.location.hash = '#/classes';
		return;
	}

	try {
		const classroom = await fetchClassById(classId);
		renderEditClassView(appElement, classroom);
	} catch (error) {
		showToast('error', error.message || 'Failed to load class details.');
		renderErrorState(appElement, error);
	}
}

function renderEditClassView(appElement, classroom) {
	const pageContent = `
		<div class="dashboard-container class-form-page">
			<section class="card class-form-card">
				<div class="class-form-intro">
					<h2>Edit Class</h2>
					<p class="text-muted">Update the details for ${safeText(classroom.class_name || 'this class')}.</p>
				</div>
			</section>

			<div class="class-form-grid">
				<section class="card class-form-card">
					<form id="edit-class-form">
						<div class="form-group">
							<label class="form-label" for="class-name-input">Class Name</label>
							<input id="class-name-input" name="class_name" class="form-input" type="text" value="${safeAttr(classroom.class_name || '')}" required />
						</div>

						<div class="form-group">
							<label class="form-label" for="grade-age-group-input">Grade / Age Group</label>
							<input id="grade-age-group-input" name="grade_age_group" class="form-input" type="text" value="${safeAttr(classroom.grade_age_group || '')}" required />
						</div>

						<div class="form-group">
							<label class="form-label">Schedule Days</label>
							<p class="text-muted text-sm" style="margin-bottom: 12px;">Pick at least one day the class meets.</p>
							<div class="schedule-days-grid">
								${renderScheduleOptions(classroom.schedule_days)}
							</div>
						</div>

						<div class="form-group">
							<label class="form-label" for="class-description-input">Description</label>
							<textarea id="class-description-input" name="description" class="form-textarea" rows="5">${safeText(classroom.description || '')}</textarea>
						</div>

						<p id="class-form-status" class="text-muted class-form-status" aria-live="polite"></p>

						<div class="class-form-actions">
							<a href="#/classes/${classroom.id}" class="btn btn-secondary">Cancel</a>
							<button id="submit-class-btn" class="btn btn-primary" type="submit">
								<i data-lucide="save"></i>
								Save Changes
							</button>
						</div>
					</form>
				</section>

				<aside class="class-form-side">
					<section class="card class-form-card">
						<h3 style="margin-bottom: 12px;">Editing tips</h3>
						<ul class="class-form-tips">
							<li>Keep the class name recognizable for teachers.</li>
							<li>Update the schedule if the meeting pattern changes.</li>
							<li>Descriptions can stay short and practical.</li>
						</ul>
					</section>
				</aside>
			</div>
		</div>
	`;

	appElement.innerHTML = createShell(pageContent, { topbarTitle: 'Edit Class' });
	attachShellHandlers();
	bindEditClassForm(classroom.id);

	if (window.lucide && typeof window.lucide.createIcons === 'function') {
		window.lucide.createIcons();
	}
}

function bindEditClassForm(classId) {
	const form = document.getElementById('edit-class-form');
	const submitButton = document.getElementById('submit-class-btn');
	const statusMessage = document.getElementById('class-form-status');
	if (!form || !submitButton || !statusMessage) return;

	const setStatusMessage = (message, visible = true) => {
		statusMessage.textContent = message;
		statusMessage.style.display = visible ? 'block' : 'none';
	};

	const setSubmitting = (isSubmitting) => {
		submitButton.disabled = isSubmitting;
		submitButton.innerHTML = isSubmitting
			? '<i data-lucide="loader-2"></i>Saving...'
			: '<i data-lucide="save"></i>Save Changes';
		if (window.lucide && typeof window.lucide.createIcons === 'function') {
			window.lucide.createIcons();
		}
	};

	form.addEventListener('submit', async (event) => {
		event.preventDefault();

		const formData = new FormData(form);
		const className = String(formData.get('class_name') || '').trim();
		const gradeAgeGroup = String(formData.get('grade_age_group') || '').trim();
		const description = String(formData.get('description') || '').trim();
		const scheduleDays = Array.from(form.querySelectorAll('input[name="schedule_days"]:checked')).map((input) => input.value);

		if (!className || !gradeAgeGroup) {
			setStatusMessage('Please complete the class name and grade / age group fields.');
			return;
		}

		if (!scheduleDays.length) {
			setStatusMessage('Please select at least one schedule day.');
			return;
		}

		setSubmitting(true);
		setStatusMessage('Saving changes...', true);

		try {
			await updateClass(classId, {
				class_name: className,
				grade_age_group: gradeAgeGroup,
				schedule_days: scheduleDays,
				description: description || undefined,
			});

			showToast('success', 'Class updated successfully.');
			window.location.hash = `#/classes/${classId}`;
		} catch (error) {
			setSubmitting(false);
			setStatusMessage(error.message || 'Failed to update class.');
			showToast('error', error.message || 'Failed to update class.');
		}
	});
}

function renderScheduleOptions(selectedDays) {
	const days = Array.isArray(selectedDays) ? selectedDays : [];
	return DAY_OPTIONS.map((day) => `
		<label class="schedule-day-option">
			<input type="checkbox" name="schedule_days" value="${safeAttr(day)}" ${days.includes(day) ? 'checked' : ''} />
			<span>${safeText(day)}</span>
		</label>
	`).join('');
}

function renderErrorState(appElement, error) {
	const pageContent = `
		<div class="dashboard-container">
			<section class="card classes-empty-state">
				<i data-lucide="alert-triangle"></i>
				<h3>Unable to load class</h3>
				<p class="text-muted">${safeText(error.message || 'Please try again in a moment.')}</p>
				<a href="#/classes" class="btn btn-secondary">Back to Classes</a>
			</section>
		</div>
	`;

	appElement.innerHTML = createShell(pageContent, { topbarTitle: 'Edit Class' });
	attachShellHandlers();

	if (window.lucide && typeof window.lucide.createIcons === 'function') {
		window.lucide.createIcons();
	}
}

function safeText(value) {
	const text = value == null ? '' : String(value);
	return text
		.replaceAll('&', '&amp;')
		.replaceAll('<', '&lt;')
		.replaceAll('>', '&gt;')
		.replaceAll('"', '&quot;')
		.replaceAll("'", '&#39;');
}

function safeAttr(value) {
	return safeText(value).replaceAll('\n', ' ');
}